docs/src/main/asciidoc/_elytron/Elytron_WildFly_Java_Security_Manager.adoc

```diff
@@ -0,0 +1,324 @@
+
+This is where doPrivileged comes in, this is effectively saying "I have verified this is safe, forget the protection domains that called me."
+So when a deployment calls into a server module and that server module calls doPrivileged then the deployments module will not longer be a part of the permission check.
+
+Adding a doPrivileged seems an obvious solution but when added we also need to think about how deployments could abuse this to get their protection domain dropped from the permissions check.
+
+==== Privileged blocks in dependencies
+Adding the privileged block in a dependency used by WildFly works because the protection domain of the server modules has all the permissions I mean, is it because the current protection domain where the dependency is has all permissions granted, or is it because it is running on a privileged block?
+
+One point to start with there is nothing special about a {oracle-javadoc}/java.base/java/security/PrivilegedAction.html[PrivilegedAction]: PrivilegedAction is an interface which allows a {oracle-javadoc}/java.base/java/lang/Runnable.html[Runnable] class to be passed in and also has a return type. If this was added later it could have been implemented with the functional interfaces. The special part is the doPrivileged call.
+Under normal circumstances every jar would be represented by it's own protection domain and each jar could be assigned it's own permissions. When the security manager performs it's permissions check it would make sure every protection domain in the call stack has been granted that permission.
+
+The WildFly modules case is slightly special as all the modules just get granted AllPermission, so we end up in a situation where deployments have a defined set of permissions and our modules have all - by default the security manager checks the permissions of both of these during a permission check that spans them - our module will of course always pass.
+
+Because of our model it is easy to get into the assumption that it is the doPrivileged call which is doing something special to bypass the security manager permissions check, it is not - all it is doing is dropping the protection domains from the call stack prior to that point, so we just end up with the protection domains for our modules on the call stack which have the all permission.
+
+So where you say the current protection domain where the dependency is has all permissions granted - yes that is it - the remaining protections domains on the call stack have the appropriate permissions granted.
+
+When I read about it, I had always been understanding that the doPriviledge does some magical things that allowed your code run as trusted code that can bypass the permissions, for example at https://docs.oracle.com/en/java/javase/11/security/java-se-platform-security-architecture.html#GUID-73F600BE-8098-4613-AD4B-E2DEFB9118D8[What It Means to Have Privileged Code], with sentences as:
```

Comment:
@boris-unckel Thank you for the updates! 

Just a minor, but would you be able to change slighty the style of these last chapters? Since this was taken from zulip it reads more like a conversation instead of documentation. WDYT?

Reply:
Ok, I assume you mean to switch from "our"/"we"... to a passive form. I was afraid my English leads to worse result but you're right - it must be a documentation style wording.

Reply:
Yes, documentation style wording would be best. Also removing the question/answer style would be good. So taht the documentation is compact. Thank you!

Reply:
@Skyllarr Done, hopefully correct

Reply:
@boris-unckel Great, thank you! It reads much better now! I just added some remaining suggestions